ruby/ql/lib/codeql/ruby/frameworks/StandardLibrary.qll

```diff
@@ -1513,53 +1489,484 @@ module Enumerable {
+      // parameter (first iteration only). All other elements are passed to the
+      // second block parameter.
       (
-        input = "ArrayElement of Receiver" and
+        input = "ArrayElement[0] of Receiver" and
         output = "Parameter[0] of BlockArgument"
         or
-        input = "ReturnValue of BlockArgument" and
-        output = "ArrayElement[?] of ReturnValue"
+        exists(ArrayIndex i | i > 0 | input = "ArrayElement[" + i + "] of Receiver") and
+        output = "Parameter[1] of BlockArgument"
       ) and
       preservesValue = true
     }
   }
 
-  private class GrepNoBlockSummary extends GrepSummary {
-    GrepNoBlockSummary() { this = "grep(no_block)" and not exists(mc.getBlock()) }
+  private class InjectArgSummary extends InjectSummary {
+    InjectArgSummary() { this = "inject(arg)" and mc.getNumberOfArguments() > 0 }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      (
+        // The first argument of the call is passed to the first block parameter.
+        input = "Argument[0]" and
+        output = "Parameter[0] of BlockArgument"
+        or
+        // Each element in the receiver is passed to the second block parameter.
+        exists(ArrayIndex i | input = "ArrayElement[" + i + "] of Receiver") and
+        output = "Parameter[1] of BlockArgument"
+      ) and
+      preservesValue = true
+    }
+  }
+
+  abstract private class MinOrMaxBySummary extends SummarizedCallable {
+    MethodCall mc;
+
+    bindingset[this]
+    MinOrMaxBySummary() { mc.getMethodName() = ["min_by", "max_by"] }
+
+    override MethodCall getACall() { result = mc }
+  }
+
+  private class MinOrMaxByNoArgSummary extends MinOrMaxBySummary {
+    MinOrMaxByNoArgSummary() {
+      this = "min_or_max_by_no_arg" and
+      mc.getNumberOfArguments() = 0
+    }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "ArrayElement of Receiver" and
+      output = ["Parameter[0] of BlockArgument", "ReturnValue"] and
+      preservesValue = true
+    }
+  }
+
+  private class MinOrMaxByArgSummary extends MinOrMaxBySummary {
+    MinOrMaxByArgSummary() {
+      this = "min_or_max_by_arg" and
+      mc.getNumberOfArguments() > 0
+    }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "ArrayElement of Receiver" and
+      output = ["Parameter[0] of BlockArgument", "ArrayElement[?] of ReturnValue"] and
+      preservesValue = true
+    }
+  }
+
+  abstract private class MinOrMaxSummary extends SummarizedCallable {
+    MethodCall mc;
+
+    bindingset[this]
+    MinOrMaxSummary() { mc.getMethodName() = ["min", "max"] }
+
+    override MethodCall getACall() { result = mc }
+  }
+
+  private class MinOrMaxNoArgNoBlockSummary extends MinOrMaxSummary {
+    MinOrMaxNoArgNoBlockSummary() {
+      this = "min_or_max_no_arg_no_block" and
```

Comment:
I think I would prefer `this = mc.getMethodName() + "_no_arg_no_block`. This way we will synthesize two different methods for calls to `min` and `max`.

Reply:
This comment may apply to other classes as well.

Reply:
Sure, I've done that for this and other classes that covered multiple methods.